enterprise/models.py

```diff
@@ -2285,34 +2341,49 @@ def enterprise_customer_user(self):
 
     def revoke(self):
         """
-        Marks this object as revoked and marks the associated EnterpriseCourseEnrollment
-        as "saved for later".  This object and the associated EnterpriseCourseEnrollment are both saved.
+        Idempotently unenroll/revoke this fulfillment and associated EnterpriseCourseEnrollment.
 
-        Subclasses may override this function to additionally emit revocation events.
+        This method and EnterpriseCourseEnrollment.set_unenrolled() call each other!!! If you edit either method, make
+        sure to preserve base cases that terminate infinite recursion.
 
-        TODO: revoke entitlements as well?
-        """
-        if self.enterprise_course_enrollment:
-            self.enterprise_course_enrollment.saved_for_later = True
-            self.enterprise_course_enrollment.unenrolled = True
-            self.enterprise_course_enrollment.unenrolled_at = localized_utcnow()
-            self.enterprise_course_enrollment.save()
+        Notes:
+        * This object and the associated EnterpriseCourseEnrollment may both be saved.
+        * Subclasses may override this function to additionally emit revocation events.
 
-        self.is_revoked = True
-        self.save()
+        Returns:
+            bool: True if self.is_revoked was changed.
+        """
+        changed = False
+        if not self.is_revoked:
+            LOGGER.info(f"Marking fulfillment {str(self)} as revoked.")
+            changed = True
+            self.is_revoked = True
+            self.save()
+            # Find and unenroll any related EnterpriseCourseEnrollment.
+            # By only updating the related object on updates to self, we prevent infinite recursion.
+            if ece := self.enterprise_course_enrollment:
```

Comment:
I wonder if it's worth short-circuiting here if the ECE is already unenrolled? Like would this be clearer/easier to grok - a reader of the code wouldn't have to jump into the logic of `set_unenrolled()` to understand how that this _will_ actually terminate:
```python
if ece := self.enterprise_course_enrollment and not ece.unenrolled:
    ece.set_unenrolled(True)
```